### PR TITLE
To support other InputMethod Implementation by python

### PR DESCRIPTION
**Pull request by kilikkuo**
Hi,
I knew this project from Taipei.py event (July30) and I'm trying to provide a machinery that which you're framework may need.
Not quit sure about if I understand your code correctly because I have no WINDOWS platform to build & run.  I can only trace server.py and try to figure it out what you may need (To be able to retrieve different IM implementation when server is launched ?). 

So, I did the following modification. Feel free to discuss with me (kilik.kuo@gmail.com), or merge this PR to your code and that would be cool : )

1. Re-org the folder structure. (python-related)
    |- server -
                   |- __init__.py
                   |- server.py
                   |- input_method -
                                               |-__init__.py
                                               |-textService.py           # Base class to inherite
                                               |-serviceManager.py  # Enumerate services and management.
                                               |-im -                             # Other installed IM implementation
                                                      |- meow -
                                                                     |- __init__.py
                                                                     |- ime_meow.py  #  Move DemoTextService here
                                                      |- woof -
                                                                   |- __init__.py
                                                                   |- woof.py
2. Add WoofTextService as another sample 
  - Like DemoTextService. 

3. Add IMServiceMgr to enumerate installed input method for client/server usage. 
 - To enumerate all installed services, and track if service is hooked to Client.
    (IIUC, 1 client holds 1 text service, right ?)

4. Add 2 attributes in JSON for IMServiceMgr instantiating service instance.
 - Need "moduleName" & "serviceName" for python dynamically import module.
 - "moduleName" stands for the name of a file under a folder which includes __init__.py (so that python could treat this folder as module)
 - "serviceName" stands for the TextService class name, used as constructor to instantiate target service.

### CODE REVIEW

**Comment by PCMan** [comment]
Thank you very much and this is just awesome! There is one thing, however, to mention.
Actually, we can use "one" text service per client. Different input methods can be treated as different "language profiles" of the same text service in Windows TSF. When switched to a different "language profile", the client receives "onLanguageProfileChanged" event with the guid of the new profile.
So, we don't really need to maintain the names of the services. They can be recognized by their GUIDs.

**Comment by kilikkuo** [comment]
Aha !   So you mean, one Client binds to one XXTextService, and XXTextService may contains several Input Methods which are defined in ime.json and these IMs could be identified separately by GUIDs.
In that case,  Client and TextService should have a API "onLanguageProfileChanged" and a GUID as input parameters for XXTextService to switch the IM, right.  But currently, I can't find "onLanguageProfileChanged", am I misunderstood ?

> So, we don't really need to maintain the names of the services. They can be recognized by their GUIDs.
To clear the fog of my confusion, the name of services here, do you mean the code |self.serviceName = "DemoText"| in class DemoTextService(ime.meow.py), or the |"serviceName" : "DemoTextService"| in ime.json ?
The front is used to set into Threading.Thread to identify what service is used by specific ClientThread. It's just to provide information. (Could be removed)
But I think the later should be kept to obtain class name in string type to instantiate the service dynamically.  The original naming is bad : (, I should probably use |"serviceClassName" : "DemoTextService"| instead.

Thoughts ?

**Comment by PCMan** [comment]
Sorry about the confusion. It's caused by bad naming of classes and the poor design of the original code. Ideally, one client only binds to one instance of TextService. In the TextService object, you can switch between different "input method engines" which can be identified by their GUIDs.
What I mean here is, DemoTextService in fact should be DemoEngine instead or something alike. It's just a an input method engine that can be loaded by TextService. Every sub directory in the input_methods folder contains an "engine" whose properties are defined in its ime.json file.
So actually DemoEngine should be derived from ImEngine, not TextService.
That's my original plan. The current code is just a quick hack so it's in a mess.

**Comment by PCMan** [comment]
Sorry about the confusion. I try to explain this better with an example. Every sub directory in the input_methods folder contains a text service whose properties are defined in its ime.json file.

# base class of all kinds of text services
class TextService:
    ....
    def onKeyDown(self, keyEvent):
    .....

class TextServiceManager:

    def __init__(self):
         self.services = dict()
         for each subdir in "input_methods" dir that contains "ime.json" file:
            json = parse_ime_json_file()
            self.services[json.guid] = <the info of the text service>

    def load(self, guid):
        if guid in self.services:
            ....

class Client:
    def __init__(self):
        self.currentTextService = ....

    def handleEvent(self, msg, guid):
        if msg is "languageProfileChanged":
            newTextService = self.serviceManager.load(guid)
            self.currentTextService.onDeactivate()
            self.currentTextService = newTextService
            newTextService.onActivate()

In short, for a client, only one type of text service is in use at a time. Different input methods are implemented as different kind of text services. The Client can switch among them when receiving the onLanguageProfileChanged message.

**Comment by kilikkuo** [comment]
Cool ~ Thanks for the explanation !!

~~Before the explanation, I thought that you'd like to create ClientThreads as much as the installed TextServices~~
So basically, now there's should be just 1 ClientThread (with 1 Client who is using specific OOTextService retrieved from IMServiceManager) communicating to server at a time.  And the client is able to switch to different TextService, say GGTextService, when client receives "onLangProfileActivate" and "onLangProfileDeactivate" msg .

I think I call just rename IMServiceManager to TextServiceManager, and add several methods like findService/ loadService/ offloadService ... etc.
And that should match what you wanna do, what do you think ?

**Comment by PCMan** [comment]
Just send some more detailed explanation (in Chinese) to your gmail. Please check your mailbox. Thanks!
